### PR TITLE
[`StableLm`] Add QK normalization and Parallel Residual Support

### PR DESCRIPTION
**Pull request by jon-tow**
# What does this PR do?

This PR adds support for the latest Stable LM 2 architecture that requires QK normalization and parallel residual layers.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

cc @ArthurZucker

### CODE REVIEW

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 1]
```diff
@@ -203,6 +203,21 @@ def forward(self, x):
         return self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
 
 
+class StableLmLayerNormPerHead(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:206]: this seems pretty similar to the `Persimmon` or `Phi` layer norm no? 

https://github.com/huggingface/transformers/blob/bbd26b001ad1f86a94129769cc61ebcaac416a3e/src/transformers/models/phi/modeling_phi.py#L273-L279

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 2]
```diff
@@ -250,6 +265,11 @@ def __init__(self, config: StableLmConfig, layer_idx: Optional[int] = None):
         self.v_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=config.use_qkv_bias)
         self.o_proj = nn.Linear(self.hidden_size, self.hidden_size, bias=False)
 
+        self.use_qk_layernorm = config.use_qk_layernorm
+        if self.use_qk_layernorm:
+            self.q_norm = StableLmLayerNormPerHead(self.head_dim, self.num_heads, eps=config.layer_norm_eps)
+            self.k_norm = StableLmLayerNormPerHead(self.head_dim, self.num_key_value_heads, eps=config.layer_norm_eps)
+
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:272]: we usually add this in a modeling code!

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 3]
```diff
@@ -719,23 +754,29 @@ def forward(
...
+            mlp_output = self.mlp(hidden_states)
+            mlp_output = self.dropout(mlp_output)
+            hidden_states = residual + self_attn_output + mlp_output
+        else:
+            # x = x + attn(ln1(x))
+            # x = x + mlp(ln2(x))
+            residual = residual + self_attn_output
+            # Fully Connected
+            mlp_output = self.mlp(self.post_attention_layernorm(residual))
+            mlp_output = self.dropout(mlp_output)
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:778]: that is also very much not transformers philosophy! 
We usually create a new architecture for that. It is tedious but we'd rather have the explicit architecture!

Comment (jon-tow) [reply src/transformers/models/stablelm/modeling_stablelm.py:778]: Gotcha - I was following the GPT NeoX convention:
https://github.com/huggingface/transformers/blob/f01e1609bf4dba146d1347c1368c8c49df8636f6/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L703-L716

I suppose I'll create a new architecture then in a separate PR and close this?

Comment (ArthurZucker) [reply src/transformers/models/stablelm/modeling_stablelm.py:778]: Ah.... if there is a precedent that might be alright

**Review by ArthurZucker — COMMENTED** [review]
There is a precedent for both so it's a bit hard for me to decline. Just use the PersimmonAttention for the layer norm , and add a comment on top of the `use_parallel_residual` to say it copied from gptNeox

File: src/transformers/models/stablelm/modeling_stablelm.py (hunk 1 above)
Comment (jon-tow) [reply src/transformers/models/stablelm/modeling_stablelm.py:206]: No. This implementation of QK normalization is per-head, i.e., each head has its own independent learnable gamma (bias unused), whereas the phi/persimmon implementations are per-layer, sharing affine parameters across all heads.

File: src/transformers/models/stablelm/modeling_stablelm.py (hunk 2 above)
Comment (jon-tow) [reply src/transformers/models/stablelm/modeling_stablelm.py]: Can you clarify, please?

File: src/transformers/models/stablelm/modeling_stablelm.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/stablelm/modeling_stablelm.py:206]: Thanks for explaining! Let's add this to the doc of this class! 😉 
Also does that mean the weights are shared for all layers? Or it's per head per layer?

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 4]
```diff
@@ -203,6 +203,21 @@ def forward(self, x):
...
+        self.dim = dim
+        self.num_heads = num_heads
+        self.norms = nn.ModuleList([nn.LayerNorm(dim, eps=eps, bias=bias) for _ in range(self.num_heads)])
+
+    def forward(self, x: torch.Tensor):
+        # Split along the num_heads axis to get per-head inputs
+        # [batch_size, num_heads, seq_len, head_dim] -> [batch_size, 1, seq_len, head_dim] * num_heads
+        heads = torch.split(x, 1, dim=1)
+        # Normalize and merge the heads back together
+        return torch.cat([norm(x) for norm, x in zip(self.norms, heads)], dim=1)
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:218]:
```suggestion
        states_per_heads = torch.split(hidden_states, 1, dim=1)
        # Normalize and merge the heads back together
        return torch.cat([norm(hidden_states) for norm, hidden_states in zip(self.norms, states_per_heads)], dim=1)
```

**Review by ArthurZucker — APPROVED** [review]
LGTM! Thanks for iterating

**Comment by ArthurZucker** [comment]
Could you add 1 generation test! 🤗

**Comment by jon-tow** [comment]
>  Could you add 1 generation test! 🤗

Added! I've further trained the tiny random test model ([stabilityai/tiny-random-stablelm-2](https://huggingface.co/stabilityai/tiny-random-stablelm-2)) on a few million tokens of [cosmopedia](https://huggingface.co/datasets/HuggingFaceTB/cosmopedia) so that failed checks are easily detectable as opposed to verifying against jumbled sample tokens/text.

Let me know if you need anything else. Thanks for the review 🙏
